add Typescript launch, function call, delphi interfaces and indexed props

smeshch opened the pull request:
<!--
Thank you for your pull request. Please review below requirements.

Bug fixes and new features should include tests and possibly benchmarks.

Contributors guide: https://github.com/nodejs/node/blob/master/CONTRIBUTING.md
-->
##### Checklist

<!-- Remove items that do not apply. For completed items, change [ ] to [x]. -->
- [ ] `make -j4 test` (UNIX), or `vcbuild test nosign` (Windows) passes
- [ ] tests and/or benchmarks are included
- [ ] documentation is changed or added
- [ ] commit message follows commit guidelines
##### Affected core subsystem(s)

<!-- Provide affected core subsystem(s) (like doc, cluster, crypto, etc). -->
##### Description of change

<!-- Provide a description of the change below this comment. -->

…ay only) add method for calling js functions

smeshch committed:
feat(): add functions to create new values (string, int, bool and array only) add method for calling js functions

smeshch committed:
chore(): rename IValue container

smeshch committed:
change some values from persistent to Local

smeshch committed:
Feat(): half-made function call, (need to split node procedures for a future)

quanterion commented on src/delphi_intf.cpp:
```diff
@@ -35,6 +35,11 @@ namespace Bazis {
+	{
+		_exe_name = name;
+	}
+
```
надо копировать в std::string

quanterion commented on src/delphi_intf.cpp:
```diff
@@ -181,25 +186,22 @@ char * IEngine::RunFileWithExePath(char * fName, char * exeName)
-char * IEngine::RunOneMoreFile(char * fName)
+char * IEngine::RunIncludeFile(char * fName)
 {
 	char * ExePath = &*(_exe_name.begin());
```
тут лучше _exe_name.data() или _exe_name.c_str()

smeshch committed:
chore(): some refactoring, clear unused variables

smeshch committed:
Fix(): fix problem with incorrect running 'include' files

smeshch committed:
chore(): some refactoring

smeshch committed:
Feat(): add node error to log, fixed string encoding/decoding between delphi & cpp

smeshch committed:
feat(): correct error log to delphi engine, correct encoding, clear memory leaks

smeshch committed:
feat(Delphi) add interfaces and their methods and props,
  add helper's props,
  add enums,
  'AddIncludeCode' method - 'code' parameter will run before running main script,
  add attributes for callback records.
feat(cpp-library):
  add check if value is undefined
feat(node-src):
  split 'StartNodeInstance' method into 'StartNodeInstance' and 'StopNodeInstance',
  change result of 'GuessHandleType' function, for working in non-console app.

smeshch committed:
add deleting engine info from isolate at engine's destructor, add check if engine exists

smeshch committed:
fix compile error

smeshch committed:
add throwing exceptions if property of object doesn't exist (e.g. for freed objects)

smeshch committed:
just local commit (trying to wrap node's main methods into class)

smeshch committed:
another local commit

smeshch committed:
add calling functions from delphi

smeshch committed:
local commit: scripts run

smeshch committed:
Fix destruction of isolate, can run two or more engines at one time

smeshch committed:
Feat(): add indexed props and cache of js objects with equals external items
Chore(): bug fix, clear comments

smeshch committed:
feat(node-original): add typescript to project, add '-f' parameter for filename;
feat(delphi): add indexed props to delphi files;
fix(): bug fix

smeshch committed:
Feat(delphi): add debug port option, delete excess info in error log output

smeshch committed:
Feat(): set to default static variables on script finish, write javascript code and source maps into file if content is typescript
